src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -88,7 +105,10 @@ export class ControllerPreferredService implements IControllerPreferredService,
     }> {
         traceInfoIfCI(`Clear controller mapping for ${getDisplayPath(document.uri)}`);
         // Keep track of a token per document so that we can cancel the search if the doc is closed
+        this.preferredCancelTokens.get(document)?.cancel();
+        this.preferredCancelTokens.get(document)?.dispose();
         const preferredSearchToken = new CancellationTokenSource();
+        this.disposables.add(preferredSearchToken);
```

Comment:
This could get called a lot for every document, over & over when new controllers are added
As a result we could end up with a lot of CancellationTokens, yes we can add them into the global Disposables list, however that means we end up slowing the main dispose method of VS Code just processing 100s if not 1000s of these when the extension is unloaded or the like.

Hence all we do here is, we keep a set of the disposables and clean up as and when they're done with, this way we end up with a significantly smaller list of disposables that need to be disposed when the extenion is disposed (in fact we could end up with just 0 being contributed by this code)